File: .agents/skills/autobot/scripts/autobot.py
import argparse
import json
import sys
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parents[4]
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

from src.core.engine.autobot_skill import execute_autobot


def main() -> None:
    parser = argparse.ArgumentParser(description="Authoritative AutoBot skill entrypoint.")
    parser.add_argument("--project-root")
    parser.add_argument("--bead-id")
    parser.add_argument("--claim-next", action="store_true")
    parser.add_argument("--checker-shell")
    parser.add_argument("--max-attempts", type=int, default=3)
    parser.add_argument("--timeout", type=float, default=300.0)
    parser.add_argument("--startup-timeout", type=float, default=30.0)
    parser.add_argument("--checker-timeout", type=float, default=300.0)
    parser.add_argument("--grace-seconds", type=float, default=3.0)
    parser.add_argument("--agent-id", default="AUTOBOT")
    parser.add_argument("--worker-note")
    parser.add_argument("--autobot-dir", default="/home/morderith/Corvus/AutoBot")
    parser.add_argument("--command", default="hermes")
    parser.add_argument("--command-arg", action="append", default=[])
    parser.add_argument("--env", action="append", default=[])
    parser.add_argument("--ready-regex", default=r"(?:^|\n)\s*❯\s*$")
    parser.add_argument("--done-regex", action="append", default=[])
    parser.add_argument(
        "--stream",
        action="store_true",
        help="Relay Hermes and checker output live. Disabled by default so stdout stays valid JSON.",
    )
    args = parser.parse_args()

    if args.claim_next and args.bead_id:
        parser.error("Use either --bead-id or --claim-next, not both.")
    if not args.claim_next and not args.bead_id:
        parser.error("Provide --bead-id or use --claim-next.")

    env = {}
    for item in args.env:
        key, sep, value = item.partition("=")
        if not sep or not key:
            parser.error(f"Invalid --env entry {item!r}; expected KEY=VALUE.")
        env[key] = value

    target_root = Path(args.project_root).resolve() if args.project_root else PROJECT_ROOT

    try:
        result = execute_autobot(
            target_root,
            bead_id=args.bead_id,
            claim_next=args.claim_next,
            autobot_dir=args.autobot_dir,
            command=args.command,
            command_args=args.command_arg,
            env=env,
            ready_regex=args.ready_regex,
            done_regexes=args.done_regex,
            timeout=args.timeout,
            startup_timeout=args.startup_timeout,
            grace_seconds=args.grace_seconds,
            no_stream=not args.stream,
            agent_id=args.agent_id,
            max_attempts=args.max_attempts,
            checker_shell=args.checker_shell,
            checker_timeout=args.checker_timeout,
            worker_note=args.worker_note,
        )
    except Exception as exc:
        print(
            json.dumps(
                {
                    "skill_id": "autobot",
                    "status": "FAILURE",
                    "outcome": "ERROR",
                    "summary": str(exc),
                    "error_type": type(exc).__name__,
                },
                indent=2,
            )
        )
        raise SystemExit(1)

    print(json.dumps(result.to_dict(), indent=2))
    raise SystemExit(0 if result.status == "SUCCESS" else 4)


if __name__ == "__main__":
    main()

File: src/node/core/runtime/weaves/autobot.ts
import fs from 'node:fs';
import path from 'node:path';

import { execa } from 'execa';

import type {
    AutobotWeaveMetadata,
    AutobotWeavePayload,
    RuntimeAdapter,
    RuntimeContext,
    WeaveInvocation,
    WeaveResult,
} from '../contracts.ts';
import { RUNTIME_KERNEL_ROOT } from '../kernel_root.ts';

function resolvePythonPath(projectRoot: string): string {
    const windows = path.join(projectRoot, '.venv', 'Scripts', 'python.exe');
    const unix = path.join(projectRoot, '.venv', 'bin', 'python');
    if (process.platform === 'win32' && fs.existsSync(windows)) {
        return windows;
    }
    if (process.platform !== 'win32' && fs.existsSync(unix)) {
        return unix;
    }
    return process.platform === 'win32' ? 'python' : 'python3';
}

function extractJsonObject(raw: string): Record<string, unknown> {
    const start = raw.indexOf('{');
    const end = raw.lastIndexOf('}');
    if (start === -1 || end === -1) {
        throw new Error('AutoBot weave did not return a JSON payload.');
    }
    return JSON.parse(raw.slice(start, end + 1)) as Record<string, unknown>;
}

export class AutoBotWeave implements RuntimeAdapter<AutobotWeavePayload> {
    public readonly id = 'weave:autobot';
    private readonly runner: typeof execa;

    public constructor(runner: typeof execa = execa) {
        this.runner = runner;
    }

    public async execute(
        invocation: WeaveInvocation<AutobotWeavePayload>,
        context: RuntimeContext,
    ): Promise<WeaveResult> {
        const payload = invocation.payload;
        if (!payload.bead_id && !payload.claim_next) {
            return {
                weave_id: this.id,
                status: 'FAILURE',
                output: '',
                error: 'AutoBot execution requires a bead_id or claim_next=true.',
            };
        }

        const kernelRoot = RUNTIME_KERNEL_ROOT;
        const targetRoot = context.workspace_root;
        const scriptPath = path.join(kernelRoot, '.agents', 'skills', 'autobot', 'scripts', 'autobot.py');
        const args = [scriptPath, '--project-root', targetRoot];

        if (payload.bead_id) {
            args.push('--bead-id', payload.bead_id);
        } else {
            args.push('--claim-next');
        }
        if (payload.checker_shell) {
            // THE GAUNTLET: Wrap the checker shell in a phantom sync and fast-failing syntax checker
            const targetPath = context.workspace_root; 
            const syncScript = path.join(RUNTIME_KERNEL_ROOT, 'src', 'node', 'core', 'runtime', 'sync_slice.ts');
            const syncCommand = `npx tsx ${syncScript} "${targetPath}" "${payload.bead_id}"`;
            
            const isTs = fs.existsSync(path.join(targetPath, 'tsconfig.json'));
            const syntaxCheck = isTs ? `npx tsc --noEmit && ` : ``;
            
            args.push('--checker-shell', `${syncCommand} && ${syntaxCheck}${payload.checker_shell}`);
        }
        if (payload.max_attempts !== undefined) {
            args.push('--max-attempts', String(payload.max_attempts));
        }
        if (payload.timeout !== undefined) {
            args.push('--timeout', String(payload.timeout));
        }
        if (payload.startup_timeout !== undefined) {
            args.push('--startup-timeout', String(payload.startup_timeout));
        }
        if (payload.checker_timeout !== undefined) {
            args.push('--checker-timeout', String(payload.checker_timeout));
        }
        if (payload.grace_seconds !== undefined) {
            args.push('--grace-seconds', String(payload.grace_seconds));
        }
        if (payload.agent_id) {
            args.push('--agent-id', payload.agent_id);
        }
        if (payload.worker_note) {
            args.push('--worker-note', payload.worker_note);
        }
        if (payload.autobot_dir) {
            args.push('--autobot-dir', payload.autobot_dir);
        }
        if (payload.command) {
            args.push('--command', payload.command);
        }
        for (const commandArg of payload.command_args ?? []) {
            args.push('--command-arg', commandArg);
        }
        if (payload.ready_regex) {
            args.push('--ready-regex', payload.ready_regex);
        }
        for (const doneRegex of payload.done_regexes ?? []) {
            args.push('--done-regex', doneRegex);
        }
        for (const [key, value] of Object.entries(payload.env ?? {})) {
            args.push('--env', `${key}=${value}`);
        }
        if (payload.stream) {
            args.push('--stream');
        }

        const { stdout } = await this.runner(resolvePythonPath(kernelRoot), args, {
            cwd: kernelRoot,
            env: { ...context.env, PYTHONPATH: kernelRoot },
        });
        const result = extractJsonObject(stdout) as AutobotWeaveMetadata & {
            status?: string;
            summary?: string;
        };

        return {
            weave_id: this.id,
            status: String(result.status) === 'SUCCESS' ? 'SUCCESS' : 'FAILURE',
            output: String(result.summary ?? 'AutoBot execution completed.'),
            error: String(result.status) === 'SUCCESS' ? undefined : String(result.summary ?? 'AutoBot execution failed.'),
            metadata: result,
        };
    }
}
